README.md
Serina | GitHub page [here](https://foxdog05000.github.io/serina/)
======

[![Build Status](https://travis-ci.org/foxdog05000/serina.svg?branch=master)](https://travis-ci.org/foxdog05000/serina)
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://github.com/foxdog05000/serina/blob/master/LICENSE)
[![Built with Grunt](https://cdn.gruntjs.com/builtwith.svg)](https://gruntjs.com/)

> GUI for create translation files for i18next

## 🖼 Preview (_Serina R0.2.3_)

![Serina Screenshot Home](https://raw.githubusercontent.com/foxdog05000/serina/master/samples/home.jpg)
![Serina Screenshot Edit translations](https://raw.githubusercontent.com/foxdog05000/serina/master/samples/edit-translation.jpg)
![Serina Screenshot View json](https://raw.githubusercontent.com/foxdog05000/serina/master/samples/view-json.jpg)
![Serina Screenshot Settings](https://raw.githubusercontent.com/foxdog05000/serina/master/samples/settings.jpg)

## 💾 Download

Serina can be downloaded [here](https://github.com/foxdog05000/serina/releases). Available versions for Linux and Windows (64bits). _MacOS available to later_

To use, unzip in the desired location and execute Serina.

If you need a different version, please follow the _development_ instructions to make your own build from the source code

## 📖 Usage
Serina allows you to create translations for each language, you can access the different languages by clicking on their map in the languages list.

### Create translation for a language
This page displays the list of languages and allows you to add a new language via a form using the ISO 639-1 language code. Example:"FR" or "EN".

The display of the languages is available in two versions (card list or list) that can be set on the parameter page. The number of elements in a language is displayed under the name of the language.

### Edit translation

This page displays the groups and translations of the first level of the translation and then recursively displays the sub-levels via access to a group allowing access to its content.

Adding a group is done via a modal that allows you to enter the group name. The modification will also be done via a modal which will ask for the new name of the group. Deleting a group will ask for confirmation before deleting.

The addition of a translation is done via a card added to the list of translations. The recording and deletion is done directly with the icons on the map.

Changing a translation triggers a key or value change detection system to know whether to display the save button in order to have a good interface.

#### Search

A search bar represented by a magnifying glass is integrated in the toolbar. The research focuses on the "keys" of translations.
The first corresponding element will be selected to highlight it. And if several items match the search, a navigation between the corresponding items is displayed next to the search bar.

Search is close if you navigate or delete a translation.

#### Parallel edition of two languages

Parallel entry of two languages is done via the "swap" (on the right-hand side of the groups and translations of the first level) comprising:
the main language (the one selected in the list of languages)
the second language that can be chosen from the other available languages.

Then it is enough to validate and the second language is automatically added below each value of the main language. Parallel editing works exactly like simple language editing.

### Parameters

The settings allow you to modify the app according to your preferences. A "About" section is also present to get details about the application.

#### Basic parameters
- The format of the language list (in card list or list)
- The language of the application (EN, ES or FR for the moment)

#### Advanced parameters
- Custom location of translation files (Available in version 0.3)

#### Others parameters
- The theme of the application
- Keyboard shortcuts

## ⁉️ FAQ

#### Your language is not available in the app ?
You can translate the application and submit your translation in PR. Or create issue.

#### Your language doesn't have its flag displayed ?
You can open a issue to explain your problem with a screenshot if possible.

## 🔨 Development

If you want to contribute to Serina, create a new different version of Serina, or compile it from source code, follow these instructions.

Clone/download source code from the github repo. Make sure you have _node_, _npm_, _yarn_ already installed in your system:

1. `yarn` to install all the dependencies.
2. `yarn start` to start _API_ on `localhost:7777`.
3. `grunt serve` to start _Serina_ on `localhost:9000`.
4. `yarn run package` to build _serina_ for linux (build will be placed in `packages` folder).
    * `grunt package` to make builds for windows, linux and mac os.
    * `grunt package:win32` to make build for windows (64bits).
    * `grunt package:linux` to make build for linux.
    * `grunt package:macos` to make build for mac os (64bits).
5. `grunt test` to run the automated tests.
    * `grunt test:loop` for loop automated tests.
    * Please, ensure the tests are passing before creating a pull requests. Add tests for your changes.

## 👏Acknowledgments

* [Electron](https://electron.atom.io)  framework was used for the app development.

## 📖 License

Serina is being developed and maintained as Open-Source software by @foxdog05000 (https://github.com/foxdog05000) licensed under [MIT LICENSE](https://github.com/foxdog05000/serina/blob/master/LICENSE)

## 📝 Source Code
The original source code can be found at: <https://github.com/foxdog05000/serina>


CHANGELOG.md
# R0.3 Comming soon

- Add preview for json files
- Migration from Bower to Yarn
- Convert commons directives to components
- Added more shortcuts in package.json
- Maj dependencies :
  - Electron version (1.7.9 -> 1.8.7)
  - AngularJS version (1.6.8 -> 1.7.5)
- Fix : Problem with count translation return 0 when the files are not empty


# [R0.2.3](https://github.com/foxdog05000/serina/releases/tag/R0.2.3) (2018-05-17)

- Fix : Bad translation "noLanguage" with "s"
- Fix : Match items on search isn't be update after delete a translation #29
- Fix : Error opening app on Mac OS #24


# [R0.2.2](https://github.com/foxdog05000/serina/releases/tag/R0.2.2) (2018-02-16)


- Added code for the language swap button
- Added rel="noopener" on links \_blank for security and performance
- Update Electron version for security reason (1.7.9 -> 1.8.2)
- Fix : Language swap button isn't working
- Fix : Bad name of file HubController in index.html


# [R0.2.1](https://github.com/foxdog05000/serina/releases/tag/R0.2.1) (2017-11-25)

- Fix : Navigate on home page or open group = search bar closed
- Fix : Buton go to top not constantly display
- Prevent multiple instances in Electron
- Maj ui swap (remove overflow)
- Remove external link on wiki page language
- Added keyboard shortcuts for toggle search bar


# [R0.2](https://github.com/foxdog05000/serina/releases/tag/R0.2) (2017-11-19)

- Adding second language selection (swap)
- CRUD improvement of groups and translations on the api side in order to manage several languages
- Improvement of the translation and group management process in the application
- Ui improvement to manage multiple languages
- Add sortJson() for sort ASC json after CRUD
- Improved download of the json file in order to keep the file formatting
- Localize theme color on settings
- Add a home button
- Add search bar on toolbar (and keyboard shortcuts)
- Fix : Toast does not support page scrolling
- Toasts are now displayed at the bottom of the screen.
- Added keyboard shortcuts list on settings page
- Ability to duplicate a translation
- Update Grunt 0.4.5 -> 1.0.0


# [R0.1.1](https://github.com/foxdog05000/serina/releases/tag/R0.1.1) (2017-09-22)

- Move button add group + translation on header of list
- Add button go to top
- Maj Electron version 1.6.11 -> 1.7.9
- Review responsive mode of app
- Button "return" : truly return precedent level
- Toast animation conflit with overflow


# [R0.1](https://github.com/foxdog05000/serina/releases/tag/R0.1) / (2017-08-31)

- Electron package
- Basic UI with Angular Material integration
- Angular 1.5 framework
- Create translation and export json file
- Change basic settings


Gruntfile.js
// Generated on 2017-03-16 using generator-angular 0.16.0
'use strict';

// # Globbing
// for performance reasons we're only matching one level down:
// 'test/spec/{,*/}*.js'
// use this if you want to recursively match all subfolders:
// 'test/spec/**/*.js'

module.exports = function (grunt) {

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);
  var serveStatic = require('serve-static');

  // Automatically load required Grunt tasks
  require('jit-grunt')(grunt, {
    useminPrepare: 'grunt-usemin',
    ngtemplates: 'grunt-angular-templates',
    cdnify: 'grunt-google-cdn'
  });

  // Configurable paths for the application
  var appConfig = {
    app: 'app',
    dist: 'dist'
  };

  // Define the configuration for all the tasks
  grunt.initConfig({

    // Project settings
    yeoman: appConfig,

    // Watches files for changes and runs tasks based on the changed files
    watch: {
      js: {
        files: ['<%= yeoman.app %>/scripts/{,*/}*.js'],
        tasks: ['newer:jshint:all', 'newer:jscs:all'],
        options: {
          livereload: '<%= connect.options.livereload %>'
        }
      },
      jsTest: {
        files: ['test/spec/{,*/}*.js'],
        tasks: ['newer:jshint:test', 'newer:jscs:test', 'karma']
      },
      compass: {
        files: ['<%= yeoman.app %>/views/{,*/}*.{scss,sass}'],
        tasks: ['compass:server', 'postcss:server']
      },
      gruntfile: {
        files: ['Gruntfile.js']
      },
      livereload: {
        options: {
          livereload: '<%= connect.options.livereload %>'
        },
        files: [
          '<%= yeoman.app %>/{,**/}*.html',
          '.tmp/views/{,*/}*.css',
          '<%= yeoman.app %>/images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}'
        ]
      }
    },

    // The actual grunt server settings
    connect: {
      options: {
        port: 9000,
        // Change this to '0.0.0.0' to access the server from outside.
        hostname: 'localhost',
        livereload: 35729
      },
      livereload: {
        options: {
          open: true,
          middleware: function (connect) {
            return [
              serveStatic('.tmp'),
              connect().use(
                '/node_modules',
                serveStatic('./node_modules')
              ),
              connect().use(
                '/app/views',
                serveStatic('./app/views')
              ),
              serveStatic(appConfig.app)
            ];
          }
        }
      },
      test: {
        options: {
          port: 9001,
          middleware: function (connect) {
            return [
              serveStatic('.tmp'),
              serveStatic('test'),
              connect().use(
                '/node_modules',
                serveStatic('./node_modules')
              ),
              serveStatic(appConfig.app)
            ];
          }
        }
      },
      dist: {
        options: {
          open: true,
          base: '<%= yeoman.dist %>'
        }
      }
    },

    // Make sure there are no obvious mistakes
    jshint: {
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      },
      all: {
        src: [
          'Gruntfile.js',
          '<%= yeoman.app %>/scripts/{,*/}*.js'
        ]
      },
      test: {
        options: {
          jshintrc: 'test/.jshintrc'
        },
        src: ['test/spec/{,*/}*.js']
      }
    },

    // Make sure code styles are up to par
    jscs: {
      options: {
        config: '.jscsrc',
        verbose: true
      },
      all: {
        src: [
          'Gruntfile.js',
          '<%= yeoman.app %>/scripts/{,*/}*.js'
        ]
      },
      test: {
        src: ['test/spec/{,*/}*.js']
      }
    },

    // Empties folders to start fresh
    clean: {
      dist: {
        files: [{
          dot: true,
          src: [
            '.tmp',
            '<%= yeoman.dist %>/{,*/}*',
            '!<%= yeoman.dist %>/.git{,*/}*'
          ]
        }]
      },
      server: '.tmp',
      package: {
        dot: true,
        src: 'packages/*'
      }
    },

    // Add vendor prefixed styles
    postcss: {
      options: {
        processors: [
          require('autoprefixer')({browsers: ['last 1 version']})
        ]
      },
      server: {
        options: {
          map: true
        },
        files: [{
          expand: true,
          cwd: '.tmp/views/',
          src: '{,*/}*.css',
          dest: '.tmp/views/'
        }]
      },
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/views/',
          src: '{,*/}*.css',
          dest: '.tmp/views/'
        }]
      }
    },

    // Compiles Sass to CSS and generates necessary files if requested
    compass: {
      options: {
        sassDir: '<%= yeoman.app %>/views',
        cssDir: '.tmp/views',
        generatedImagesDir: '.tmp/images/generated',
        imagesDir: '<%= yeoman.app %>/images',
        javascriptsDir: '<%= yeoman.app %>/scripts',
        fontsDir: '<%= yeoman.app %>/views/fonts',
        httpImagesPath: '/images',
        httpGeneratedImagesPath: '/images/generated',
        httpFontsPath: '/views/fonts',
        relativeAssets: false,
        assetCacheBuster: false,
        raw: 'Sass::Script::Number.precision = 10\n'
      },
      dist: {
        options: {
          generatedImagesDir: '<%= yeoman.dist %>/images/generated'
        }
      },
      server: {
        options: {
          sourcemap: true
        }
      }
    },

    // Renames files for browser caching purposes
    filerev: {
      dist: {
        src: [
          '<%= yeoman.dist %>/scripts/{,*/}*.js',
          '<%= yeoman.dist %>/views/{,*/}*.css',
          '<%= yeoman.dist %>/images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}',
          '!<%= yeoman.dist %>/images/flags/*',
          '<%= yeoman.dist %>/views/fonts/*'
        ]
      }
    },

    // Reads HTML for usemin blocks to enable smart builds that automatically
    // concat, minify and revision files. Creates configurations in memory so
    // additional tasks can operate on them
    useminPrepare: {
      html: '<%= yeoman.app %>/index.html',
      options: {
        dest: '<%= yeoman.dist %>',
        flow: {
          html: {
            steps: {
              js: ['concat', 'uglifyjs'],
              css: ['cssmin']
            },
            post: {}
          }
        }
      }
    },

    // Performs rewrites based on filerev and the useminPrepare configuration
    usemin: {
      html: ['<%= yeoman.dist %>/{,**/}*.html'],
      css: ['<%= yeoman.dist %>/views/{,*/}*.css'],
      js: ['<%= yeoman.dist %>/scripts/{,*/}*.js'],
      options: {
        assetsDirs: [
          '<%= yeoman.dist %>',
          '<%= yeoman.dist %>/images',
          '<%= yeoman.dist %>/views'
        ],
        patterns: {
          js: [[/(images\/[^''""]*\.(png|jpg|jpeg|gif|webp|svg))/g, 'Replacing references to images']]
        }
      }
    },

    // The following *-min tasks will produce minified files in the dist folder
    // By default, your `index.html`'s <!-- Usemin block --> will take care of
    // minification. These next options are pre-configured if you do not wish
    // to use the Usemin blocks.
    // cssmin: {
    //   dist: {
    //     files: {
    //       '<%= yeoman.dist %>/views/main.css': [
    //         '.tmp/views/{,*/}*.css'
    //       ]
    //     }
    //   }
    // },
    // uglify: {
    //   dist: {
    //     files: {
    //       '<%= yeoman.dist %>/scripts/scripts.js': [
    //         '<%= yeoman.dist %>/scripts/scripts.js'
    //       ]
    //     }
    //   }
    // },
    concat: {
      dist: {
        files: {
          'dist/main.js': ['api/api.js', '<%= yeoman.app %>/main.js']
        }
      }
    },

    imagemin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/images',
          src: '{,*/}*.{png,jpg,jpeg,gif,ico}',
          dest: '<%= yeoman.dist %>/images'
        }]
      }
    },

    svgmin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/images',
          src: '{,*/}*.svg',
          dest: '<%= yeoman.dist %>/images'
        }]
      }
    },

    htmlmin: {
      dist: {
        options: {
          collapseWhitespace: true,
          conservativeCollapse: true,
          collapseBooleanAttributes: true,
          removeCommentsFromCDATA: true
        },
        files: [{
          expand: true,
          cwd: '<%= yeoman.dist %>',
          src: ['*.html'],
          dest: '<%= yeoman.dist %>'
        }]
      }
    },

    ngtemplates: {
      dist: {
        options: {
          module: 'serinaApp',
          htmlmin: '<%= htmlmin.dist.options %>',
          usemin: 'scripts/scripts.js'
        },
        cwd: '<%= yeoman.app %>',
        src: 'views/{,**/}*.html',
        dest: '.tmp/templateCache.js'
      }
    },

    // ng-annotate tries to make the code safe for minification automatically
    // by using the Angular long form for dependency injection.
    ngAnnotate: {
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/concat/scripts',
          src: '*.js',
          dest: '.tmp/concat/scripts'
        }]
      }
    },

    // Replace Google CDN references
    cdnify: {
      dist: {
        html: ['<%= yeoman.dist %>/*.html']
      }
    },

    // Copies remaining files to places other tasks can use
    copy: {
      dist: {
        files: [{
          expand: true,
          dot: true,
          cwd: '<%= yeoman.app %>',
          dest: '<%= yeoman.dist %>',
          src: [
            '*.{ico,png,txt}',
            '*.html',
            'images/{,*/}*.{webp}',
            'icons/icon-x64.ico',
            'views/fonts/{,*/}*.*',
            'package.json'
          ]
        }, {
          expand: true,
          cwd: '.tmp/images',
          dest: '<%= yeoman.dist %>/images',
          src: ['generated/*']
        }, {
          expand: true,
          cwd: 'app/app/',
          dest: '<%= yeoman.dist %>',
          src: '**'
        }, {
          expand: true,
          cwd: 'api/',
          dest: '<%= yeoman.dist %>',
          src: ['modules/*']
        }, {
          expand: true,
          cwd: 'node_modules/',
          dest: '<%= yeoman.dist %>/node_modules',
          src: [
            'accepts/**',
            'array-flatten/**',
            'body-parser/**',
            'bytes/**',
            'content-disposition/**',
            'content-type/**',
            'cookie-signature/**',
            'debug/**',
            'depd/**',
            'destroy/**',
            'ee-first/**',
            'encodeurl/**',
            'escape-html/**',
            'express/**',
            'finalhandler/**',
            'forwarded/**',
            'iconv-lite/**',
            'inherits/**',
            'ipaddr.js/**',
            'jsonfile/**',
            'http-errors/**',
            'etag/**',
            'fresh/**',
            'cookie/**',
            'media-typer/**',
            'merge-descriptors/**',
            'methods/**',
            'mime/**',
            'mime-db/**',
            'mime-types/**',
            'ms/**',
            'negotiator/**',
            'on-finished/**',
            'parseurl/**',
            'path-to-regexp/**',
            'proxy-addr/**',
            'qs/**',
            'range-parser/**',
            'raw-body/**',
            'send/**',
            'safe-buffer/**',
            'safer-buffer/**',
            'serve-static/**',
            'setprototypeof/**',
            'statuses/**',
            'type-is/**',
            'unpipe/**',
            'utils-merge/**',
            'vary/**'
          ]
        }, {
          expand: true,
          cwd: 'node_modules/material-design-icons/iconfont',
          dest: '<%= yeoman.dist %>/styles',
          src: [
            'MaterialIcons-Regular.ttf',
            'MaterialIcons-Regular.woff',
            'MaterialIcons-Regular.woff2'
          ]
        }]
      },
      styles: {
        expand: true,
        cwd: '<%= yeoman.app %>/views',
        dest: '.tmp/views/',
        src: '{,*/}*.css'
      }
    },

    // Run some tasks in parallel to speed up the build process
    concurrent: {
      server: [
        'compass:server'
      ],
      test: [
        'compass'
      ],
      dist: [
        'compass:dist',
        'imagemin',
        'svgmin'
      ]
    },

    // Build app electron
    electron: {
      buildWin32: {
        options: {
          name: 'serina',
          dir: 'dist',
          out: 'packages',
          platform: 'win32',
          arch: 'x64',
          overwrite: true,
          prune: false
        }
      },
      buildLinux: {
        options: {
          name: 'serina',
          dir: 'dist',
          out: 'packages',
          platform: 'linux',
          arch: 'x64',
          overwrite: true,
          prune: false
        }
      },
      buildMacos: {
        options: {
          name: 'serina',
          dir: 'dist',
          out: 'packages',
          platform: 'darwin',
          arch: 'x64',
          overwrite: true,
          prune: false
        }
      }
    },

    // Test settings
    karma: {
      unit: {
        configFile: 'test/karma.conf.js',
        singleRun: true
      },
      loop: {
        configFile: 'test/karma.conf.js',
        singleRun: false
      }
    }
  });


  grunt.registerTask('serve', 'Compile then start a connect web server', function (target) {
    if (target === 'dist') {
      return grunt.task.run(['build', 'connect:dist:keepalive']);
    }

    grunt.task.run([
      'test',
      'clean:server',
      'concurrent:server',
      'postcss:server',
      'connect:livereload',
      'watch'
    ]);
  });

  grunt.registerTask('server', 'DEPRECATED TASK. Use the "serve" task instead', function (target) {
    grunt.log.warn('The `server` task has been deprecated. Use `grunt serve` to start a server.');
    grunt.task.run(['serve:' + target]);
  });

  grunt.registerTask('test', 'Execute TU', function (target) {
    target = (target === 'loop') ? ':' + target : ':unit';

    grunt.task.run([
      'clean:server',
      'concurrent:test',
      'postcss',
      'connect:test',
      'karma' + target
    ]);
  });

  grunt.registerTask('build', [
    'test',
    'clean:dist',
    'clean:package',
    'useminPrepare',
    'concurrent:dist',
    'postcss',
    'ngtemplates',
    'concat',
    'ngAnnotate',
    'copy:dist',
    'cssmin',
    'uglify',
    'filerev',
    'usemin',
    'htmlmin'
  ]);

  grunt.registerTask('package', 'Package electron app', function (target) {
    switch (target) {
      case 'linux':
        grunt.task.run([
          'build',
          'electron:buildLinux'
        ]);
        break;
      case 'win32':
        grunt.task.run([
          'build',
          'electron:buildWin32'
        ]);
        break;
      case 'macos':
        grunt.task.run([
          'build',
          'electron:buildMacos'
        ]);
        break;
      default:
        grunt.task.run([
          'build',
          'electron:buildLinux',
          'electron:buildWin32',
          'electron:buildMacos'
        ]);
    }
  });

  grunt.registerTask('default', [
    'newer:jshint',
    'newer:jscs',
    'test',
    'build'
  ]);
};
